java-checks/src/main/java/org/sonar/java/checks/NullPointerCheck.java

```diff
@@ -0,0 +1,147 @@
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.ArrayAccessExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import javax.annotation.Nullable;
+
+@Rule(
+  key = "S2259",
+  name = "Null pointers should not be dereferenced",
+  tags = {"bug", "cert", "cwe", "owasp-a1", "owasp-a2", "owasp-a6", "security"},
+  priority = Priority.BLOCKER)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.ERRORS)
+@SqaleConstantRemediation("20min")
+public class NullPointerCheck implements JavaFileScanner {
+
+  public static final String KEY = "S2259";
+  private static final RuleKey RULE_KEY = RuleKey.of(CheckList.REPOSITORY_KEY, KEY);
+
+  private JavaFileScannerContext context;
+  private ExpressionVisitor expressionVisitor = new ExpressionVisitor();
+  private IssueVisitor issueVisitor = new IssueVisitor();
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    if (context.getSemanticModel() != null) {
+      expressionVisitor.scan(context.getTree());
+    }
+  }
+
+  /** this visitor invokes issueVisitor for the base expressions of array access, member selectm and method invocation expressions. */
+  private class ExpressionVisitor extends BaseTreeVisitor {
+    @Override
+    public void scan(@Nullable Tree tree) {
+      super.scan(tree);
+    }
+
+    @Override
+    public void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+      issueVisitor.scan(tree.expression());
+      super.visitArrayAccessExpression(tree);
+    }
+
+    @Override
+    public void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+      issueVisitor.scan(tree.expression());
+      super.visitMemberSelectExpression(tree);
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      scan(tree.typeArguments());
+      for (ExpressionTree argument : tree.arguments()) {
+        issueVisitor.scan(argument);
+      }
+      scan(tree.methodSelect());
+    }
+  }
+
+  /** this visitor raises an issue if it visits an identifier, method invocation or member select whose symbol can be null. */
+  private class IssueVisitor extends BaseTreeVisitor {
+    @Override
+    public void scan(@Nullable Tree tree) {
+      super.scan(tree);
+    }
+
+    @Override
+    public void visitIdentifier(IdentifierTree tree) {
+      Symbol symbol = ((SemanticModel) context.getSemanticModel()).getReference((IdentifierTree) tree);
```

Comment:
SemanticModel should probably be a field of the class updated by the scanFile method to avoid multiple casting to it.